Zend/zend_partial.c

```diff
@@ -0,0 +1,932 @@
+		zend_hash_add(ht, info->name, value);
+	} else {
+		zend_internal_arg_info *internal = (zend_internal_arg_info*) info;
+
+		zend_hash_str_add(ht, internal->name, strlen(internal->name), value);
+	}
+}
+
+static zend_always_inline void zend_partial_debug_fill(zend_partial *partial, HashTable *ht) {
+	zval *arg = partial->argv,
+		 *aend = arg + partial->argc;
+	zend_arg_info *info = partial->func.common.arg_info,
+				  *iend = info + partial->func.common.num_args;
+
+	while (info < iend) {
+		zval param;
+
+		ZVAL_NULL(&param);
+
+		if (arg < aend) {
+			if (Z_IS_NOT_PLACEHOLDER_P(arg)) {
+				ZVAL_COPY_VALUE(&param, arg);
+			}
+
+			arg++;
+		}
+
+		zend_partial_debug_add(&partial->func, ht, info, &param);
+
+		if (Z_OPT_REFCOUNTED(param)) {
+			Z_ADDREF(param);
+		}
+		info++;
+	}
+
+	if (ZEND_PARTIAL_FUNC_FLAG(&partial->func, ZEND_ACC_VARIADIC)) {
+		zval variadics;
+
+		array_init(&variadics);
+
+		zend_partial_debug_add(&partial->func, ht, info, &variadics);
+
+		while (arg < aend) {
+			if (Z_IS_NOT_PLACEHOLDER_P(arg)) {
+				zend_hash_next_index_insert(Z_ARRVAL(variadics), arg);
+
+				if (Z_OPT_REFCOUNTED_P(arg)) {
+					Z_ADDREF_P(arg);
+				}
+			}
+
+			arg++;
+		}
+
+		if (partial->named) {
+			zend_hash_merge(Z_ARRVAL(variadics), partial->named, zval_copy_ctor, true);
+		}
+	}
+}
+
+static HashTable *zend_partial_debug(zend_object *object, int *is_temp) {
+	zend_partial *partial = (zend_partial*) object;
+	zval args;
+	HashTable *ht;
+
+	ht = zend_closure_get_debug_info(object, is_temp);
+
+	array_init(&args);
+	zend_hash_update(ht, ZSTR_KNOWN(ZEND_STR_ARGS), &args);
+
+	zend_partial_debug_fill(partial, Z_ARRVAL(args));
+
+	return ht;
+}
+
+static HashTable *zend_partial_get_gc(zend_object *obj, zval **table, int *n)
+{
+	zend_partial *partial = (zend_partial *)obj;
+
+	if (!partial->argc) {
+		*table = Z_TYPE(partial->This) == IS_OBJECT ? &partial->This : NULL;
+		*n = Z_TYPE(partial->This) == IS_OBJECT ? 1 : 0;
+    } else {
+		zend_get_gc_buffer *buffer = zend_get_gc_buffer_create();
+
+		if (Z_TYPE(partial->This) == IS_OBJECT) {
+			zend_get_gc_buffer_add_zval(buffer, &partial->This);
+		}
+
+		for (uint32_t arg = 0; arg < partial->argc; arg++) {
+			zend_get_gc_buffer_add_zval(buffer, &partial->argv[arg]);
+		}
+
+		zend_get_gc_buffer_use(buffer, table, n);
+    }
+
+	return NULL;
+}
+
+static zend_function *zend_partial_get_method(zend_object **object, zend_string *method, const zval *key) /* {{{ */
+{
+	if (zend_string_equals_literal_ci(method, ZEND_INVOKE_FUNC_NAME)) {
+		zend_partial *partial = (zend_partial*) *object;
+
+		return &partial->trampoline;
+	}
+
+	return zend_std_get_method(object, method, key);
+}
+/* }}} */
+
+static int zend_partial_get_closure(zend_object *obj, zend_class_entry **ce_ptr, zend_function **fptr_ptr, zend_object **obj_ptr, bool check_only)
+{
+	zend_partial *partial = (zend_partial*) obj;
+
+	*fptr_ptr = &partial->trampoline;
+	*obj_ptr  = (zend_object*) &partial->std;
```

Comment:
It would be good to initialize ce_ptr as well here. It looks like call sites don't use it if obj_ptr is set, but I wouldn't rely on it.

Reply:
Fixed in 6afa0d104ec0a6d46b5fc6a96521336306c3cd1f